Implemented CommonTimeBlock

MMachadoUofT opened the pull request:
I definitely see why TDD is so highly regarded... lots of tiny bugs I wouldn't have been able to find otherwise. I made sure to include thorough tests to assure that this suite was as penetrative as possible. The rest of the remaining classes are very simple, as they are basically just containers for the ones I've created thus far, and any of their functionality will primarily involve calling their sub-entities' methods.

MMachado05 committed:
Created the TimeBlockFactory classes

This includes one abstract, and one concrete.

MMachado05 committed:
Implemented CommonTimeBlockFactory.create()

This included writing the attributes and constructor for the CommonTimeBlock class.

MMachado05 committed:
Implemented CommonTimeBlock equals and hashCode methods

The latter method was overridden to continue adhering to the equals() contract.

MMachado05 committed:
Implemented CommonTimeBlock comparison methods

- overlapsWith
- isContainedWithin
- contains
- isAdjacentTo

MMachado05 committed:
Implemented getters/setters, getDuration

MMachado05 committed:
Created CommonTimeBlock test class

MMachado05 committed:
Merge branch 'main' into 13-subissue-15-timeblockclasses

MMachado05 committed:
Created methods for testing equals()

This includes an implementation for the null comparison test.

MMachado05 committed:
Created test methods for TimeBlock.overlapsWith()

MMachado05 committed:
Created test methods for TimeBlock.isContainedWithin()

MMachado05 committed:
Created test methods for TimeBlock.isAdjacentTo()

MMachado05 committed:
Created test methods for TimeBlock.getDuration()

I've also added a small note to CommonTimeBlock's constructor referring to the possibility of client code providing invalid start and end times, where the former comes *after* the latter.

MMachado05 committed:
Implemented TimeBlock.equals() tests

I've also added a new one testing that adjacent blocks are not equal.

MMachado05 committed:
Implemented TimeBlock.overlapsWith() tests

This included fixing a small bag and adding specificity to the interface javadoc.

MMachado05 committed:
Implemented TimeBlock.isContainedWithin() tests

This included fixing the method to better fit the documentation.

MMachado05 committed:
Implemented TimeBlock.isContainedWithin() tests

This included fixing the method to better fit the documentation.

MMachado05 committed:
Implemented TimeBlock.isAdjacentTo() tests

This included fixing a bug where I compared object addresses rather than using the equals() method.

MMachado05 committed:
Implemented TimeBlock.getDuration() tests

This included changing the implementation to consistently return the correct value.

dadler83 commented on src/main/java/entity_layer/CommonTimeBlock.java:
```diff
@@ -11,6 +12,26 @@
+    public CommonTimeBlock(LocalDateTime startTime, LocalDateTime endTime) {
+        this.startTime = startTime;
+        this.endTime = endTime;
+        // I'll have to include a check somewhere that doesn't allow start time to be before
```
nonsensical comment

dadler83 replied:
If you are going to implement the check do so, if not, let a higher level object implement it

dadler83 commented on src/main/java/entity_layer/TimeBlock.java:
```diff
@@ -27,7 +27,8 @@ public interface TimeBlock {
     int hashCode();
```
I don't think you need hashCode here because it is overridden from the object class

dadler83 reviewed (changes requested):
If the tests all pass, it looks pretty good. I left some minor comments about cleaning up the code though, nothing functionality wise to comment on -- it looks good..

MMachado05 committed:
Comment and interface cleanup

MMachado05 committed:
Merge branch 'main' into 13-subissue-15-timeblockclasses

MMachado05 committed:
Extracted isAdjacentTo into Before and After

I've kept isAdjacentTo, as well; this extraction was made as I was planning out the code for TimeBlockManager and I realized that having dedicated methods for these cases would help, but so would the original, as well.

MMachado05 committed:
Extracted overlapsWith into Before and After

I've kept overlapsWith, as well; this extraction was made as I was planning out the code for TimeBlockManager and I realized that having dedicated methods for these cases would help, but so would the original, as well.

MMachado05 committed:
Made TimeBlock extend Serializable

This will allow for easy data persistence.

MMachado05 committed:
Merge branch 'main' into 13-subissue-15-timeblockclasses

dadler83 reviewed (approved):
These entities look good for now; we'll start using them for the use cases, and clean up code as needed.

Amir1441 reviewed (approved):
The methods look more and more complete now. Each one is more tuned into the specific role it has to play in the entirety of the project. Looks good. With this, the Time Block usecases can now run
